### PR TITLE
Add serializer to serialize webapp configurations to pom

### PR DESCRIPTION
**Pull request by Flanker32**
Add serializer to serialize webapp configurations to pom, for webapp:config

### CODE REVIEW

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/WebAppConfiguration.java [hunk 1]
```diff
@@ -280,6 +289,7 @@ public Builder filtering(final MavenResourcesFiltering value) {
             this.filtering = value;
             return self();
         }
+
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/WebAppConfiguration.java]: redundant empty line

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java [hunk 2]
```diff
@@ -41,21 +51,13 @@ public JavaVersion getJavaVersion() {
...
-                    return RuntimeStack.TOMCAT_8_5_JRE8;
-                case TOMCAT_9_0:
-                    return RuntimeStack.TOMCAT_9_0_JRE8;
-                case WILDFLY_14:
-                    return RuntimeStack.WILDFLY_14_JRE8;
-                default:
-                    throw new MojoExecutionException(
-                        String.format("Unknown value of <webContainer>. Supported values are %s, %s and %s.",
-                            TOMCAT_8_5, TOMCAT_9_0, WILDFLY_14));
+            final String fixWebContainer = StringUtils.isEmpty(webContainer) ? JRE_8 : webContainer;
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: `fixWebContainer` needs to get better naming, cuz JRE_8` is not a web container

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java [hunk 3]
```diff
@@ -69,6 +71,10 @@ public WebContainer getWebContainer() {
         return WebContainer.fromString(webContainer);
     }
 
+    public static String getLinuxJavaVersionByRuntimeStack(RuntimeStack runtimeStack) {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: Better naming. The returned string is not only Java version.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    private void configWindowsRunTime(WebAppConfiguration webAppConfiguration, Element domElement) {
+        domElement.add(XMLUtils.createSimpleElement("javaVersion",
+            webAppConfiguration.getJavaVersion().toString()));
+        domElement.add(XMLUtils.createSimpleElement("webContainer",
+            webAppConfiguration.getWebContainer().toString()));
+    }
+
+    private void configLinuxRunTime(WebAppConfiguration webAppConfiguration, Element domElement) {
+        domElement.add(XMLUtils.createSimpleElement("linuxRuntime",
+            linuxRuntimeStackMap.get(webAppConfiguration.getRuntimeStack())));
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: Can we use the result of `RuntimeStack::toString()`?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V2ConfigurationSerializer.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        XMLUtils.removeNode(configurationElement, "runtime");
+        configurationElement.add(createRunTimeNode(webAppConfiguration));
+
+        XMLUtils.removeNode(configurationElement, "deploymentSlot");
+        if (webAppConfiguration.getDeploymentSlotSetting() != null) {
+            configurationElement.add(createDeploymentSlotNode(webAppConfiguration.getDeploymentSlotSetting()));
+        }
+
+        // only add deployment if user didn't set it
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V2ConfigurationSerializer.java]: typo: `didn't`

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java (hunk 4 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: No, RuntimeStack::toString() will return result with jre version which was not supported value for our configuration  (https://github.com/Microsoft/azure-maven-plugins/blob/develop/azure-webapp-maven-plugin/README.md#details-explanation-of-some-configurations)

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java (hunk 3 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: yes, changed it to `getLinuxWebContainerByRuntimeStack`

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java (hunk 2 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: changed it to `fixRuntime`

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java (hunk 4 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: V1Configuration use RuntimeStack.toString() as linux runtime directly, so change it in next commit.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import com.microsoft.azure.maven.webapp.utils.XMLUtils;
+import org.apache.maven.plugin.MojoFailureException;
+import org.dom4j.Element;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class V1ConfigurationSerializer extends ConfigurationSerializer {
+
+    private static final Map<RuntimeStack, String> linuxRuntimeStackMap = new HashMap<>();
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: Can be removed

Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: Removed

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java [hunk 7]
```diff
@@ -41,21 +51,13 @@ public JavaVersion getJavaVersion() {
...
-                    throw new MojoExecutionException(
-                        String.format("Unknown value of <webContainer>. Supported values are %s, %s and %s.",
-                            TOMCAT_8_5, TOMCAT_9_0, WILDFLY_14));
+            final String fixRuntime = StringUtils.isEmpty(webContainer) ? JRE_8 : webContainer;
+            if (runtimeStackMap.containsKey(fixRuntime)) {
+                return runtimeStackMap.get(fixRuntime);
+            } else {
+                throw new MojoExecutionException(
+                    String.format("Unknown value of <webContainer>. Supported values are %s.",
+                        StringUtils.join(runtimeStackMap.keySet().toArray(), ",")));
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: toArray is not needed?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/ConfigurationSerializer.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+public abstract class ConfigurationSerializer {
+
+    public abstract void saveToXML(WebAppConfiguration webAppConfiguration, Element document)
+        throws MojoFailureException;
+
+    protected DOMElement createResourcesNode(List<Resource> resources) {
+        final DOMElement resourceRootNode = new DOMElement("resources");
+        for (final Resource resource : resources) {
+            final DOMElement resourceNode = new DOMElement("resource");
+            if (StringUtils.isNotEmpty(resource.getFiltering())) {
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/ConfigurationSerializer.java]: a private util method to check  and create

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/ConfigurationSerializer.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (StringUtils.isNotEmpty(resource.getFiltering())) {
+                resourceNode.add(XMLUtils.createSimpleElement("targetPath", resource.getTargetPath()));
+            }
+
+            if (StringUtils.isNotEmpty(resource.getDirectory())) {
+                resourceNode.add(XMLUtils.createSimpleElement("directory", resource.getDirectory()));
+            }
+
+            if (resource.getIncludes() != null && resource.getIncludes().size() > 0) {
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/ConfigurationSerializer.java]: a private util method to check and create

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        throws MojoFailureException {
+        createOrUpdateAttribute("schemaVersion", "V1", configurationElement);
+        createOrUpdateAttribute("resourceGroup", webAppConfiguration.getResourceGroup(), configurationElement);
+        createOrUpdateAttribute("appName", webAppConfiguration.getAppName(), configurationElement);
+        createOrUpdateAttribute("region", webAppConfiguration.getRegion().name(), configurationElement);
+        createOrUpdateAttribute("pricingTier",
+            PricingTierEnum.getPricingTierStringByPricingTierObject(webAppConfiguration.getPricingTier()),
+            configurationElement);
+
+        coinfigRuntime(webAppConfiguration, configurationElement);
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/serializer/V1ConfigurationSerializer.java]: typo coinfigRuntime

**Pull request merged**